### PR TITLE
8300235: Use VarHandle access in Image(Input | Output)StreamImpl classes

### PR DESCRIPTION
**Pull request by minborg**
This PR suggests improving performance by using the newly introduced class `jdk.internal.util.ByteArray` to improve packing/unpacking operations.

The PR also proposes adding a `ByteArrayLittleEndian` class for support for little endian packing/unpacking.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300235](https://bugs.openjdk.org/browse/JDK-8300235): Use VarHandle access in Image(Input | Output)StreamImpl classes

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [68d72946](https://git.openjdk.org/jdk/pull/12204/files/68d729463ce43bddf9fd7d98a9e7d1302ba50a1d)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12204/head:pull/12204` \
`$ git checkout pull/12204`

Update a local copy of the PR: \
`$ git checkout pull/12204` \
`$ git pull https://git.openjdk.org/jdk pull/12204/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12204`

View PR using the GUI difftool: \
`$ git pr show -t 12204`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12204.diff">https://git.openjdk.org/jdk/pull/12204.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/module-info.java [hunk 1]
```diff
@@ -272,6 +272,8 @@
         jdk.jfr;
     exports jdk.internal.util.random to
         jdk.random;
+    exports jdk.internal.util to
```
Comment (minborg) [inline src/java.base/share/classes/module-info.java:275]: Added coupling I am afraid.

**Comment by bplb** [comment]
Looks good overall. Did you measure any performance changes in the image streams? (Note that this PR and its corresponding issue are missing "Stream" in the title.)

**Comment by minborg** [comment]
Here are some performance figures:

Baseline (J21 master)
```
     [java] All test results:
     [java] Test(imageio.input.stream.tests.readFully(int[])) averaged 2.0089243986475606E9 bytes/sec
     [java]     with 250x250, photo, byteArray, !useCache
     [java] Test(imageio.input.stream.tests.readLong) averaged 6.087925258517034E8 bytes/sec
     [java]     with 250x250, photo, byteArray, !useCache
```

Patch
```
     [java] All test results:
     [java] Test(imageio.input.stream.tests.readFully(int[]) averaged 2.0537614726652384E9 bytes/sec
     [java]     with 250x250, photo, byteArray, !useCache
     [java] Test(imageio.input.stream.tests.readLong) averaged 6.227265050569915E8 bytes/sec
     [java]     with 250x250, photo, byteArray, !useCache
```

the `readFully(int[])` is a custom benchmark:

```
                iis.mark();
                int[] array = new int[8];
                do {
                    if (pos + 4*8 > length) {
                        iis.reset();
                        iis.mark();
                        pos = 0;
                    }
                    iis.readFully(array, 0, 8);
                    pos += 4*8;
                } while (--numReps >= 0);
```

**Comment by minborg** [comment]
Given the way the benchmarks are written, I suspect the actual performance gains are higher than indicated by the figures above. The benchmark likely spends more time in preparing for the call (e.g. checking, looping) than in the actual target method.

**Comment by RogerRiggs** [comment]
Can the benchmark be moved to the JMH framework; it deals with the setup and measurements in a more predictable way.
See test/micro/org/openjdk/bench/jajva/nio/... for examples.

**Comment by minborg** [comment]
> Can the benchmark be moved to the JMH framework; it deals with the setup and measurements in a more predictable way.
> See test/micro/org/openjdk/bench/jajva/nio/... for examples.

This is t

**Comment by minborg** [comment]
Here are some benchmarks:

```
Baseline (JDK21, master)
Benchmark                       Mode  Cnt  Score   Error  Units
ImageInputStreamBench.readInt   avgt   15  1.289 ± 0.029  ns/op
ImageInputStreamBench.readLong  avgt   15  1.308 ± 0.017  ns/op

Patch
Benchmark                       Mode  Cnt  Score   Error  Units
ImageInputStreamBench.readInt   avgt   15  0.946 ± 0.046  ns/op
ImageInputStreamBench.readLong  avgt   15  0.954 ± 0.010  ns/op

```

image

File: src/java.base/share/classes/jdk/internal/util/ByteArrayLittleEndian.java [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+     * @throws IndexOutOfBoundsException if the provided {@code offset} is outside
+     *                                   the range [0, array.length - 2]
+     * @see #getDoubleRaw(byte[], int)
+     */
+    public static void setDoubleRaw(byte[] array, int offset, double value) {
+        // Just sets the bits as they are
+        DOUBLE.set(array, offset, value);
+    }
+
+    private static VarHandle create(Class<?> viewArrayClass) {
```
Comment (RogerRiggs) [inline src/java.base/share/classes/jdk/internal/util/ByteArrayLittleEndian.java:420]: Renaming the method to `createLittleEndian` would make the static assignments at the top unambiguous.

**Review by RogerRiggs — APPROVED** [review]
Looks good. 
Thanks for the JMH benchmarks.

**Comment by AlanBateman** [comment]
Do I read this correctly that the proposal is to have ByteArray and ByteArrayLittleEndian?

**Pull request closed without merging**